### DISCUSSION TITLE
Root object Save with Children

### DISCUSSION THREAD

**Discussion by Art666OTS** [MarimerLLC/csla#2514] [Questions] [answered]
Hi,

Background this is for Web and API use only.

Up until now I have manually managed children when using CSLA and I thought it was time to utilize CSLA as I keep reading about saving the root is all you have to do.

My thoughts were that in the save of the root CSLA does some magic and goes through the collection of children that are Businessbase and utilizes the save on each child.

I had a look at the example SimpleNTier and it appears that this is not the case.  You need to setup some code in the Fetch and Update to tell CSLA to process the children.  I'm OK with that however when I get to the persistence in the database I keep getting an error "System.InvalidOperationException: 'Can not directly save a child object'"

I know that you cannot save a child object using .Save() as I have had this before.

The root Update is this

```c#
       #region Update
        public ArticleEdit Update(ArticleEdit obj)
        {
            if (obj.IsDeleted)
            {
                if (!obj.IsNew)
                {
                    // delete data
                    Delete(obj.ID);
                    return Create();
                }
                MarkNew(obj);
            }
            else
            {
                if (obj.IsNew)
                {
                    DoInsert(obj);
                }
                else
                {
                    DoUpdate(obj);
                }
                // Do any child objects here
                var cf = new CategoryFactory();
                cf.UpdateItems(obj, obj.ChildObj);
                MarkOld(obj);
            }
            return obj;
        }
```

and the UpdateItems code is 

```c#
       internal void UpdateItems(ArticleEdit a, CategoryItems lineItems)
        {
            if (lineItems == null)
            {
                return;
            }
            if (lineItems.Count.Equals(0))
            {
                return;
            }
            var delList = GetDeletedList<CategoryEdit>(lineItems);
            foreach (var item in delList)
            {
                //  code to delete each one.
                item.Delete(item.CAID);
            }
            delList.Clear();
            foreach (var item in lineItems)
            {
                var i = item;
                if (i.IsSavable)
                {
                    if (i.IsNew)
                    {
                        i = i.Save();
                    }
                    else
                    {
                        if (i.IsDirty)
                        {
                           i = i.Save();
                        }
                    }
                    MarkOld(i);
                }
            }
        }
```

If I change the code to this it works

```c#
       internal void UpdateItems(ArticleEdit a, CategoryItems lineItems)
        {
            if (lineItems == null)
            {
                return;
            }
            if (lineItems.Count.Equals(0))
            {
                return;
            }
            var delList = GetDeletedList<CategoryEdit>(lineItems);
            foreach (var item in delList)
            {
                //  code to delete each one.
                Csla.DataPortal.Delete<CategoryEdit>(item.CAID);
            }
            delList.Clear();
            foreach (var item in lineItems)
            {
                var i = item;
                if (i.IsSavable)
                {
                    if (i.IsNew)
                    {
                        Csla.DataPortal.Update<CategoryEdit>(item);
                    }
                    else
                    {
                        if (i.IsDirty)
                        {
                            Csla.DataPortal.Update<CategoryEdit>(item);
                        }
                    }
                    MarkOld(i);
                }
            }
        }
```

However I feel that I am going back to the client side to accomplish this. Is this the correct way to do this or is there a better way?

I have also seen code that refers to FieldManager.UpdateChildren(). Should I be using this?

Should I have picked a better example to review?

Thanks

**Comment by rockfordlhotka** [comment, marked as answer]
SimpleNTier uses the _factory implementation_ model for data access. When you use the factory model, you assume a lot of control (and responsibility) for managing the metastate of the business domain objects.

Personally, I tend to use the _encapsulated invocation_ model mostly, because it allows CSLA to manage the metastate, and still provides clean separation of concerns between the business layer and data access layer.

These models are discussed (along with the other valid models) in the [Using CSLA 4 book](https://store.lhotka.net/using-csla-4-all-books) series, specifically the book on data access.

The FieldManager.UpdateChildren supports the encapsulated invocation (and encapsulated implementation) models, and is generally the right answer.

If you choose to use a factory model, then the object factory for the root type is responsible for calling factory methods for the child objects.

In no case should the child class implement any direct data access method like you've done. Saving child objects is the responsibility of either the root object (inside the data portal operation methods like `Insert` or `Update`), or the root object factory.

In the SimpleNTier example you can see how the [Update method in OrderFactory](https://github.com/MarimerLLC/csla/blob/0eddc403ea0a32994d5bcb7e0350ed601febf4c9/Samples/SimpleNTier/DataAccess/OrderFactory.cs#L73) saves the root and all child objects. The child objects are saved by delegating to the LineItemFactory, which is responsible for persisting child line items.

Contrast that to the ProjectTracker sample, where the ProjectEdit class implements [insert and update data portal operation methods](https://github.com/MarimerLLC/csla/blob/0eddc403ea0a32994d5bcb7e0350ed601febf4c9/Samples/ProjectTracker/ProjectTracker.BusinessLibrary.Netstandard/ProjectEdit.cs#L219) that leverage the data portal to manage metastate, and that use the UpdateChildren method to have the data portal invoke the `InsertChild` and `UpdateChild` and `DeleteSelfChild` operation methods in the ProjectResourceEdit class.

**Reply by Art666OTS** [reply]
Thanks for your reply. Sorry I have taken so long to respond. It has taken me quite some time to review what I had done wrong and select the appropriate model to go forward with. There are so many interesting things in the samples its easy to get lost.

I have chosen the EncapsulatedInvokeDTO as a style and I think that this will provide me with a clean solution that I can comprehend to move to CSLA5+

While I think I have correctly implemented this I did notice that when a child has been altered and you perform a save on the parent it updates the parent and the child. Is this supposed to happen?  Everything appears fine after the save IsDirty etc are all as expected.

Thanks again for putting me back on track.

**Reply by rockfordlhotka** [reply]
> While I think I have correctly implemented this I did notice that when a child has been altered and you perform a save on the parent it updates the parent and the child. Is this supposed to happen?

Yes, when you save a root object, the entire object graph is saved. That includes the root, its children, grandchildren, etc.

**Reply by Art666OTS** [reply]
Thanks for confirming this.

I look forward to seeing new articles etc on CSLA 5+. Is there anywhere in particular I should keep my eye on?

Thanks again
Arthur
